bpf, Hubble: Add `is_reply` information (when available) at the `TO_OVERLAY` observability point

qmonnet opened the pull request:
The objective of this PR is to add the relevant information to indicate to Hubble whether an event is a reply packet or not, at the `TO_OVERLAY` observability point (for encapsulated packets).

We have the information available in most cases, but it requires some piping through the encapsulating functions. There are also a few cases where we only get the information when the host firewall is enabled.

Please refer to individual commit logs for details.

qmonnet committed:
bpf: pipe forwarding reason into traces for TO_OVERLAY

Add an additional argument to __encap_with_nodeid() so we can pass the
reason (conntrack status) for forwarding the packet when calling
send_trace_notify(). This allows us to get the "is_reply" field in the
flows visible in Hubble.

We also need to let Hubble know that the TraceToOverlay observability
point now has reply information available.

Note that the reason remains at zero at most locations where
__encap_with_nodeid() is called: only the trivial cases where we have
the info were addressed in this commit, others will be updated as a
follow-up.

In the near future, we want to clean up the confusion between an
intended 0 value for the reason (TRACE_REASON_POLICY) and an unspecified
reason. This is not addressed in the current work. However, in order to
avoid complicating things further, we introduce a new macro
TRACE_REASON_UNSPECIFIED, currently set to 0, to make it easier to
rework the reason values in the future.

Signed-off-by: Quentin Monnet <quentin@isovalent.com>

qmonnet committed:
bpf: pass reason from higher up when available for TO_OVERLAY traces

The previous commit added the possibility to pass a reason to the traces
at the TO_OVERLAY observability. The reason was effectively passed for
trivial cases. This commit adds some more plumbing to collect the reason
from higher-level functions, when available.

In some locations, we only get the reason from the conntrack lookup if
the host firewall is enabled. Retrieving this reason requires a bit more
piping through the host firewall functions, and is left for a separate
commit.

Signed-off-by: Quentin Monnet <quentin@isovalent.com>

qmonnet committed:
bpf: retrieve conntrack lookup result from host firewall (egress)

The host firewall often performs a conntrack lookup for its own needs.
When it is enabled, it may be useful to reuse the result from the lookup
for other purposed as well. This commit adds a new struct trace_ctx
container structure to pass both the monitor parameter and the
forwarding reason to ipv*_host_policy_egress(), in order to have them
populated by the conntrack lookup performed by the host firewall and to
see this reason in Hubble traces.

Co-authored-by: Daniel Borkmann <daniel@iogearbox.net>
Signed-off-by: Quentin Monnet <quentin@isovalent.com>

joestringer commented on bpf/lib/encap.h:
```diff
@@ -139,15 +139,16 @@ __encap_with_nodeid(struct __ctx_buff *ctx, __u32 tunnel_endpoint,
 
 	send_trace_notify(ctx, TRACE_TO_OVERLAY, seclabel, 0, 0, ENCAP_IFINDEX,
-			  0, monitor);
+			  (enum trace_reason)ct_reason, monitor);
```
Can the function parameter just take an `enum trace_reason`?

joestringer replied:
I only commented on this case since I saw the cast here in the first patch, but I think there's a similar question to ask for the later patches as well.

qmonnet replied:
Makes sense, thanks for catching this. I've addressed it.

joestringer reviewed (commented):
Code LGTM, it just seems like we could make more use of the dedicated enum types to get stronger guarantees about passing the correct values around. Maybe in the end it's not a big deal if the notification format requires int, but it'd be nice to avoid int parameters being passed around, particularly if there is a more specific type that we expect.

qmonnet committed:
bpf: pass reason for TO_NETDEV through host firewall

In bpf_host's to_netdev()'s call to send_trace_notify() we do not only
want to have the monitor parameter that has been populated by the host
firewall's CT lookup, but also the trace reason from it. The previous
commit gave it to us for free, let's just add the reason to the traces.

Fixes: #18460
Co-authored-by: Daniel Borkmann <daniel@iogearbox.net>
Signed-off-by: Quentin Monnet <quentin@isovalent.com>

gandro reviewed (commented):
I'm a bit worried about the remaining `TRACE_REASON_UNSPECIFIED` cases in nodeport. If they are ever called for reply traffic, it will break Hubble UI.

I haven't been able to find an example in practice however, so maybe those are never called for reply traffic?

qmonnet committed:
bpf: retrieve conntrack lookup result from host firewall (ingress)

The host firewall often performs a conntrack lookup for its own needs.
When it is enabled, it may be useful to reuse the result from the lookup
for other purposed as well. This commit uses a struct trace_ctx
container structure to pass both the monitor parameter and the
forwarding reason to ipv*_host_policy_ingress(), in order to have them
populated by the conntrack lookup performed by the host firewall and to
see this reason in Hubble traces.

Signed-off-by: Quentin Monnet <quentin@isovalent.com>

qmonnet committed:
bpf: pass a few more reasons for traces

Add a few more reasons for traces for some observability points where we
the reason is in fact a constant:

- In do_netdev_encryp_encap(), we can use TRACE_REASON_ENCRYPTED.
- In tail_handle_arp(), we know that we prepare and send an ARP
  response. Therefore, we can pass TRACE_REASON_CT_REPLY to indicate
  that the packet is an answer to a request.
- In tail_nodeport_nat_ipv{4,6}(), the dir == NAT_DIR_EGRESS branch is
  executed for N/S LB requests (as per the comment in the code). So we
  can indicate that this is _not_ a reply: let's pass CT_NEW (more
  explicit that the equivalent TRACE_REASON_POLICY).

Signed-off-by: Quentin Monnet <quentin@isovalent.com>

borkmann commented on bpf/bpf_host.c:
```diff
@@ -560,11 +563,9 @@ handle_ipv4(struct __ctx_buff *ctx, __u32 secctx,
 						     info->key, secctx,
-						     TRACE_REASON_UNKNOWN,
-						     TRACE_PAYLOAD_LEN);
+						     trace.reason, trace.monitor);
```
nit: Could we just use `struct trace_ctx` for the encap helpers?

qmonnet replied:
I did it at first for both `encap_and_redirect_with_nodeid()` and `__encap_and_redirect_with_nodeid()`, but I ended up creating a lot of `struct trace_ctx` variables to fill it with values we already had, just for the sake of passing it to the functions. Although the calls were slightly cleaner, I thought we didn't gain much in readability in the callers overall. Are you sure you want the `struct trace_ctx` everywhere?

qmonnet replied:
Addressed (for most helpers) with last commit.

borkmann reviewed (approved):
rest lgtm

gandro reviewed (approved):
Thanks for taking care of this, looks good to me!

qmonnet committed:
bpf: group trace reason and monitor into struct trace_ctx where relevant

There are several locations where grouping the reason and "monitor"
(length) for traces helps reading the code. Let's use structs to pass
them to most of the encapsulation helpers.

There are also a few helpers where it does not help much, because we
often pass direct values, and grouping with a struct would make us
declare many structs just for the sake of complying with the new
argument type. For such cases, we do not change the code to take structs
instead of the two integers. This is the case, for example, for
__encap_with_nodeid(), which is left untouched. Similarly,
send_trace_notify() is not changed to take a struct.

Suggested-by: Daniel Borkmann <daniel@iogearbox.net>
Signed-off-by: Quentin Monnet <quentin@isovalent.com>

joestringer commented on bpf/lib/nodeport.h:
```diff
@@ -1996,7 +2006,8 @@ static __always_inline int rev_nodeport_lb4(struct __ctx_buff *ctx, int *ifindex
 encap_redirect:
-	ret = __encap_with_nodeid(ctx, tunnel_endpoint, SECLABEL, TRACE_PAYLOAD_LEN);
+	ret = __encap_with_nodeid(ctx, tunnel_endpoint, SECLABEL,
+				  (enum trace_reason)ct_ret, TRACE_PAYLOAD_LEN);
```
Legitimate failure report from CI:

```
2022-03-29T04:40:34.852989973Z level=warning msg="In file included from /var/lib/cilium/bpf/bpf_overlay.c:33:" subsys=datapath-loader
2022-03-29T04:40:34.852992210Z level=warning msg="/var/lib/cilium/bpf/lib/nodeport.h:2010:26: error: variable 'ct_ret' is uninitialized when used here [-Werror,-Wuninitialized]" subsys=datapath-loader
2022-03-29T04:40:34.852994565Z level=warning msg="                                  (enum trace_reason)ct_ret, TRACE_PAYLOAD_LEN);" subsys=datapath-loader
2022-03-29T04:40:34.852996838Z level=warning msg="                                                     ^~~~~~" subsys=datapath-loader
2022-03-29T04:40:34.852999055Z level=warning msg="/var/lib/cilium/bpf/lib/nodeport.h:1872:12: note: initialize the variable 'ct_ret' to silence this warning" subsys=datapath-loader
2022-03-29T04:40:34.853001316Z level=warning msg="        int ct_ret, ret, fib_ret, ret2, l3_off = ETH_HLEN, l4_off;" subsys=datapath-loader
2022-03-29T04:40:34.853003514Z level=warning msg="                  ^" subsys=datapath-loader
2022-03-29T04:40:34.853005883Z level=warning msg="                   = 0" subsys=datapath-loader
2022-03-29T04:40:34.853008057Z level=warning msg="1 error generated." subsys=datapath-loader
2022-03-29T04:40:34.853012385Z level=fatal msg="Error while creating daemon" error="error while initializing daemon: exit status 1" subsys=daemon
```

qmonnet replied:
Yes thank you, was working on it :) There's a path with a `goto` where the variable is uninitialised indeed. Alas due to the combinations of macros it didn't show up when compiling locally. Should be fixed now!
